### PR TITLE
FEAT: can now train simsiam with CIFAR10; also added downstream task …

### PR DESCRIPTION
**Pull request by b-elfo**
…script; will need to clean it up;

### CODE REVIEW

File: data.py [hunk 1]
```diff
@@ -6,23 +6,36 @@
...
 from torchvision import transforms
+from torchvision.datasets import CIFAR10
 
 ###
 
 class SimSiamDataset(Dataset):
     def __init__(self, 
-                 path):
+                 path,
+                 size: int = 256,
```
Comment (b-elfo) [inline data.py:16]: output height/width size

File: data.py [hunk 2]
```diff
@@ -6,23 +6,36 @@
...
+        else:
+            return len(self.img_dirs)
 
     def __getitem__(self, 
                      idx):
-        img_dir = os.path.join( self.path, self.img_dirs[idx] )
-        img = Image.open(img_dir)
+        if self.path=='CIFAR10':
+            img, _ = self.dataset.__getitem__(idx)
+            # img = img.numpy()
```
Comment (b-elfo) [inline data.py:35]: Forcing CIFAR10 dataset to fit into my dataset;

File: data.py [hunk 3]
```diff
@@ -55,18 +68,22 @@ def simsiam_transforms(self,
                                    transforms.RandomApply([transforms.ColorJitter(brightness=brightness, contrast=contrast, saturation=saturation, hue=hue)], p=colour_p),
                                    transforms.RandomGrayscale(p=gray_p),
                                    transforms.RandomApply([transforms.GaussianBlur(kernel_size=kernel, sigma=sigma)], p=gauss_p),
-                                   transforms.ToTensor()
+                                   #transforms.ToTensor()
```
Comment (b-elfo) [inline data.py:71]: This is only commented out for the CIFAR dataset;

File: loss.py [hunk 4]
```diff
@@ -26,8 +26,8 @@ def forward(self,
         # z1 = z1.unsqueeze(dim=0)
         # z2 = z2.unsqueeze(dim=0)
 
-        z1 = z1.detach()
-        z2 = z2.detach()
+        # z1 = z1.detach()
+        # z2 = z2.detach()
```
Comment (b-elfo) [inline loss.py:30]: Moved to `forward()` function

File: main.py [hunk 5]
```diff
@@ -11,16 +11,23 @@
 ###
 
 def train():
-    init_lr = 5e-2
+    init_lr = 5e-3
     batch_size = 16
     shuffle = True
     num_workers = 4
-    num_epochs = 20
+    num_epochs = 30
```
Comment (b-elfo) [inline main.py:18]: Hyperparam tuning

File: main.py [hunk 6]
```diff
@@ -11,16 +11,23 @@
...
 
-    train_dataloader, valid_dataloader = dataloader('/media/elfo/Elfo Bkup/Seagate Portable/bird_identifier/train', #'/home/elfo/side_projects/simsiam/feature_extractor/birds_unlablled_noframes',
+    dataset_path = 'CIFAR10'
+    size = 32
+    # dataset_path = '/media/elfo/Elfo Bkup/Seagate Portable/bird_identifier/train' # needs to be fixed, add 'glob.glob' to dataset imag dir collection
+    # dataset_path = '/home/elfo/side_projects/simsiam/feature_extractor/birds_unlablled_noframes'
+    # size = 256
```
Comment (b-elfo) [inline main.py:24]: Previous paths for bird datasets

File: model.py [hunk 7]
```diff
@@ -75,6 +76,42 @@ def forward(self,
                 ):
         X = self.backbone(X)
         X_proj = self.projector(X)
+        X_proj = X_proj.detach()
```
Comment (b-elfo) [inline model.py:79]: For some reason having the `.detach()` here works, which was the main error stopping learning;

File: model.py [hunk 8]
```diff
@@ -75,6 +76,42 @@ def forward(self,
...
+                 device: str = 'cuda',
+                 ):
+        super().__init__()
+        self.head = nn.Linear(1000, num_classes, device=device)
+
+    def forward(self,
+                X: torch.Tensor,
+                ):
+        X = self.head(X)
+        return X
```
Comment (b-elfo) [inline model.py:117]: Down-sampling task network

File: train.py [hunk 9]
```diff
@@ -18,8 +18,6 @@ def do_train_epoch(dataloader: DataLoader,
     model = model.train()
     total_loss = 0
     step = 0
-    p1_hist, p2_hist = [], []
-    z1_hist, z2_hist = [], []
```
Comment (b-elfo) [inline train.py:22]: Variables for debugging;

File: train.py [hunk 10]
```diff
@@ -36,21 +34,25 @@ def do_train_epoch(dataloader: DataLoader,
...
-        if i%1==0:
-            wandb.log({'running_loss':total_loss, 
+        if i%100==0:
+            wandb.log({'current_loss':loss,
+                       'running_loss':total_loss, 
                        'batch_num':i, 
+                       'p1_mean':p1.mean(),
+                       'p2_mean':p2.mean(),
+                       'z1_mean':z1.mean(),
+                       'z2_mean':z2.mean(),
```
Comment (b-elfo) [inline train.py:44]: Logging more variables for debugging; may remove soon;

File: train.py [hunk 11]
```diff
@@ -36,21 +34,25 @@ def do_train_epoch(dataloader: DataLoader,
...
                        'p1_std':p1.std(dim=1).mean(),
                        'p2_std':p2.std(dim=1).mean(),
                        'z1_std':z1.std(dim=1).mean(),
                        'z2_std':z2.std(dim=1).mean(),
                        })
 
+        optim.zero_grad()
```
Comment (b-elfo) [inline train.py:51]: moving this above `optim.step`

File: train.py [hunk 12]
```diff
@@ -111,4 +114,6 @@ def do_n_epochs(train_dataloader: DataLoader,
 
         print("Epoch: {}\tTrain loss: {:.2f}\tValid loss: {:.2f}\n".format(epoch, train_loss, valid_loss))
 
-        torch.save( obj=model.state_dict(), f=f'model_{epoch+1}.pth' )
+        dataset_name = dataset_path.split('/')[-1]
+
+        torch.save( obj=model.backbone.state_dict(), f='models/'+dataset_name+'/'+f'pretrain_model_{epoch+1}.pth' )
```
Comment (b-elfo) [inline train.py:119]: adding complexity to `save` function;
